### PR TITLE
Updates to support creation of DR7 HLSPs

### PR DESCRIPTION
**Pull request by jotaylor**
Many updates, detailed below. I have tested this code and used it to create draft DR7 products, so I know it works :)

- `apply_cos_shifts.py`: if overwrite is True, remove calcos products and re-calibrate.
- new script `combine_header_keys.py`: This creates a new abstract class whose sole purpose is to combine header keywords. The reason I pulled this out into its own script was that multiple scripts used this logic and it was becoming untenable to keep updating duplicated code. The abstract class logic is the same as was used before, just in its own container.
- `coadd.py`: Have SegmentList inherit from new KeyBlender class and remove the code that was used to combine header keywords since that is in KeyBlender.
- `ullyses_hlsp.py`: Inherit from `KeyBlender` and remove combine header keyword logic. Fix a couple header keys in XSU data.
- `ullyses_coadd_abut_wrapper.py`: Again, remove the combine header keywords logic and use it as inherited from coadd.SegmentList. Fix data type checking to work with np.int data (this will be obsolete once we merge Robert's HASP abut code)
- `create_ullyses_hlsp.py`: Add a wrapper function to create new XSU HLSPs.
- `stis_coadd_x1d.py, timeseries.py`: Update API to coadd to use new `inpath` and `infiles` varnames.
- `stisblazefix_wrapper.py, unflag_echelle_orders.py`: Fix typos and logic mistakes
- Add ability to include new 1st header COMMENT to keep track of special calibration the ULLYSES team applied to each dataset.
- `timeseries_wrapper.py`: Improve, but leave unchanged, some functionality, fix the expected naming scheme of yaml files, rename serendipitous->exp & monitoring->subexp, add more sophisticated data checking to ensure correct inputs are being used.

### CODE REVIEW

**Comment by jotaylor** [comment]
@emsnyder can you review the timeseries updates?

File: ullyses/timeseries.py [hunk 1]
```diff
@@ -366,10 +368,10 @@ def process_files(grating, outfile, indir=".", wavelength_binning=1, min_exptime
     # and to contain the headers used to create the provenance table
     # Ullyses_COSSegmentLists have the Ullyses-specific get_target() and get_coords()
     # methods
-    ensemble = create_ensemble_segmentlist(grating, indir, wavelength_binning, ins=ins)
+    ensemble = create_ensemble_segmentlist(grating, indir, wavelength_binning, ins=ins, infiles=infiles)
     # If there's only 1 matching file for hte specified grating, exit 
     if len(ensemble.datasets) == 1:
-        print(f"SKIPPING: only one matching dataset for grating {grating}")
+        print(f"{RED}SKIPPING: only one matching dataset for grating {grating}{RESET}")
```
Comment (emsnyder) [inline ullyses/timeseries.py:374]: does this make the line print out red? that's cool!

File: ullyses/timeseries.py
Comment (emsnyder) [inline ullyses/timeseries.py:1]: Looks good to go

File: ullyses/timeseries_wrapper.py
Comment (emsnyder) [inline ullyses/timeseries_wrapper.py:1]: Also looks good - I like the refactor to exposure/sub-exposure away from serendipitous/monitoring star

File: ullyses/timeseries.py (hunk 1 above)
Comment (jotaylor) [reply ullyses/timeseries.py:374]: Yes! Some terminal magic

File: ullyses/coadd.py [hunk 2]
```diff
@@ -9,12 +9,12 @@
...
 
 from ullyses_utils.ullyses_config import CAL_VER
-
+from ullyses.combine_header_keys import KeyBlender
 # coadd data
 #
 
 STIS_NON_CCD_DETECTORS = ['FUV-MAMA', 'NUV-MAMA']
-class SegmentList:
+class SegmentList(KeyBlender):
```
Comment (stscirij) [inline ullyses/coadd.py:17]: This seems a bit extreme - I like to think of the SegmentList class as the class that handles coaddition of spectra, and now you're deriving it from a separate class that has functionality over all Ullyses products.  I get that you want to be able to do self.combine_keys(), but it feels like adding this to the SegmentList base class is rather too high a price to pay in that it overcomplicates what should be a simple base class. 
I'd suggest keeping the current combine_keys method in the segmentlist class, and if you want the ullyses_segmentlist class to have different, more extensive behaviour that includes FUSE, LCOGT, WFC3, VLT, have it derive from KeyBlender and SegmentList.
The rule should be that importing a SegmentList from coadd shouldn't require importing any other modules in the Ullyses package (technically, not even CAL_VER) to preserve the lack of dependency of hasp on ullyses

File: ullyses/unflag_echelle_orders.py [hunk 3]
```diff
@@ -26,11 +26,16 @@ def unflag_2048(filename):
...
+                if len(ind[0]) == 0:
+                    performed = False
+                    continue
                 hdulist[i].data["DQ"][ind[0][0]] -= np.bitwise_and(hdulist[i].data["DQ"][ind[0][0]], 2048)
         # Since custom processing was performed, mark these as level0
-        hdulist[0].header["HLSP_LVL"] = 0
-    print(f"Subtracted DQ=2048 flag from {opt_elem}/{cenwave} order={order} for {filename}")
+        if performed is True:
+            hdulist[0].header["HLSP_LVL"] = 0
+            print(f"Subtracted DQ=2048 flag from {opt_elem}/{cenwave} order={order} for {filename}")
```
Comment (stscirij) [inline ullyses/unflag_echelle_orders.py:41]: Maybe replace 'subtracted' with 'zeroed out' or 'unset' or 'reset' or something like that?

**Comment by jotaylor** [comment]
@stscirij let me know if things look okay now, with the latest changes

**Comment by stscirij** [comment]
Yes, this looks much better!

**Review by stscirij — APPROVED** [review]
Looks much better and cleaner now!

**Pull request merged**